### ISSUE TITLE
Tests glob* fails under Fedora 27, but not on macOS 10.12.6

### ISSUE THREAD

**Issue by catull** [att/ast#255]
In an unmodified workspace, as current as commit c1d427490c5d395c7d1ac965910b09d9581bf9bb, all 3 `glob` tests fail.
Ran the test suite 3 times, all other tests succeed.

```log
 58/156 glob                                    FAIL     0.05 s
 59/156 glob(C.UTF-8)                           FAIL     0.05 s
 60/156 glob(shcomp)                            FAIL     0.06 s
```

Ironically, all 3 of them succeed under macOS 10.12.6, is it already Christmas ?

The failed tests on my Fedora box display these lines:

```log
 58/156 glob                                    FAIL     0.05 s

--- command ---
SHELL='/home/catull/workspace/shell/korn/ast.git/build/src/cmd/ksh93/ksh' LANG='' LD_LIBRARY_PATH='/home/catull/workspace/shell/korn/ast.git/build/src/lib/libast:/home/catull/workspace/shell/korn/ast.git/build/src/lib/libcmd:/home/catull/workspace/shell/korn/ast.git/build/src/lib/libcoshell' /home/catull/workspace/shell/korn/ast.git/build/src/cmd/ksh93/ksh /home/catull/workspace/shell/korn/ast.git/src/cmd/ksh93/tests/glob.sh
--- stderr ---
    glob.sh[156] glob -- expected '<d> <dd> <de> <Beware>', got '<Beware> <d> <dd> <de>'
    glob.sh[276] glob -- expected '<abc> <abe> <bdir> <ca> <de> <man> <Beware>', got '<abc> <abe> <bdir> <Beware> <ca> <de> <man>'
-------

 59/156 glob(C.UTF-8)                           FAIL     0.05 s

--- command ---
SHELL='/home/catull/workspace/shell/korn/ast.git/build/src/cmd/ksh93/ksh' LANG='C.UTF-8' LD_LIBRARY_PATH='/home/catull/workspace/shell/korn/ast.git/build/src/lib/libast:/home/catull/workspace/shell/korn/ast.git/build/src/lib/libcmd:/home/catull/workspace/shell/korn/ast.git/build/src/lib/libcoshell' /home/catull/workspace/shell/korn/ast.git/build/src/cmd/ksh93/ksh /home/catull/workspace/shell/korn/ast.git/src/cmd/ksh93/tests/glob.sh
--- stderr ---
    glob.sh[156] glob -- expected '<d> <dd> <de> <Beware>', got '<Beware> <d> <dd> <de>'
    glob.sh[276] glob -- expected '<abc> <abe> <bdir> <ca> <de> <man> <Beware>', got '<abc> <abe> <bdir> <Beware> <ca> <de> <man>'
-------

 60/156 glob(shcomp)                            FAIL     0.06 s

--- command ---
SHELL='/home/catull/workspace/shell/korn/ast.git/build/src/cmd/ksh93/ksh' LANG='' SHCOMP='/home/catull/workspace/shell/korn/ast.git/build/src/cmd/ksh93/shcomp' LD_LIBRARY_PATH='/home/catull/workspace/shell/korn/ast.git/build/src/lib/libast:/home/catull/workspace/shell/korn/ast.git/build/src/lib/libcmd:/home/catull/workspace/shell/korn/ast.git/build/src/lib/libcoshell' /home/catull/workspace/shell/korn/ast.git/build/src/cmd/ksh93/ksh /home/catull/workspace/shell/korn/ast.git/src/cmd/ksh93/tests/shcomptest /home/catull/workspace/shell/korn/ast.git/src/cmd/ksh93/tests/glob.sh
--- stderr ---
    tmp.HgOLgspjmT[156] glob -- expected '<d> <dd> <de> <Beware>', got '<Beware> <d> <dd> <de>'
    tmp.HgOLgspjmT[276] glob -- expected '<abc> <abe> <bdir> <ca> <de> <man> <Beware>', got '<abc> <abe> <bdir> <Beware> <ca> <de> <man>'
-------
```

**Comment by siteshwar** [comment]
@catull We are running all the tests under fedora 27 on Travis and these tests are passing, so it may be an issue with how your system is configured. To debug the tests, you should put ```set -x``` on top of glob test script and run it as ```meson test glob -v```.

**Comment by catull** [comment]
Part of the reason there are discrepancies is due to the different CFLAGS used in meson.build and .travis.yml.

My Fedora 27 box has the following installation:

dnf install ed gcc langpacks-zh_CN meson sudo binutils cpp glibc-devel glibc-headers isl kernel-headers libgomp libmpc ninja-build vim-filesystem glibc-langpack-zh

I copied this from a successful Travis CI build.

I have filed an issue with a patch to even out the discrepancies as far as the compilation is concerned, between meson and Travis, see issue #258 and PR #259.

**Comment by catull** [comment]
`meson test glob -v` gives me still:

```log
ninja: Entering directory `/home/carlo/workspace/shell/korn/ast.git/build'
ninja: no work to do.
	glob.sh[156] glob -- expected '<d> <dd> <de> <Beware>', got '<Beware> <d> <dd> <de>'
	glob.sh[276] glob -- expected '<abc> <abe> <bdir> <ca> <de> <man> <Beware>', got '<abc> <abe> <bdir> <Beware> <ca> <de> <man>'
1/1 glob                                    FAIL     0.47 s

OK:         0
FAIL:       1
SKIP:       0
TIMEOUT:    0
```

**Comment by catull** [comment]
Found the problem:

I have a few extra env variables: LC_ALL=en_US.UTF-8.

If I unset it, or set it to 'C', the test does not fail anymore!

Closing this, as resolved.

**Comment by krader1961** [comment]
Thanks for bringing this to our attention, @catull. This is an example of something that might have been okay more than a decade ago but is no longer valid.

I'm going to reopen this because unit tests should be immune to the user's values for locale env vars. I have `LANG=en_US.UTF-8` in my environment but do not have a `LC_ALL` var. Simply setting `LC_ALL` in my environment should not affect unit test behavior. Unit tests are, by definition, supposed to be immune to the preferences of the user running them.

**Comment by catull** [comment]
As far as I can tell, `LC_ALL` is set to 'C' in the `Mamfile`s.
If so, the tests must also run with `LC_ALL`='C'.

Indeed, they would still succeed.

Perhaps the current value of LC_ALL is stored before running the tests, if there is one.
And restored at the end, if there is one to restore.

**Comment by siteshwar** [comment]
@krader1961 Is it fixed by #481 ? If not, it should be trivial to fix it now.

**Comment by catull** [comment]
Yepp, a bit late to the party, can confirm they pass now under macOS 10.12.6:

```log
....
 58/153 glob                                    OK       1.11 s
 59/153 glob/C.UTF-8                            OK       0.50 s
 60/153 glob/shcomp                             OK       0.55 s
....
```
